### PR TITLE
feat: support SAML and test with aliyun IDaaS

### PR DESCRIPTION
**Pull request by seriouszyx**
Signed-off-by: Yixiang Zhao <seriouszyx@foxmail.com>

Fix: https://github.com/casbin/casdoor/issues/140

Basically, I implemented authentication using the SAML protocol and tested it with Aliyun IDaaS. 

Click on the Aliyun icon to jump to the Aliyun IDaaS platform for login, the user `admin` in Casdoor corresponds to the username `casdoor`, the password is `Casdoor123!`. 

After logging in, you can access the resources of Casdoor.

https://user-images.githubusercontent.com/33992371/144571133-ba16bed0-8e9c-4bb8-8b08-08c2388367f7.mp4

The progress is shown in the following diagram.

SAML

The demo link is http://82.156.216.21:8000/.

In fact, there is some overlap with the OAuth protocol implementation, which I may have misunderstood. And there is still more work to be done, including logout and encryption.

If there is anything I haven't explained clearly or if you have any questions, please tell me!

### CODE REVIEW

**Comment by hsluoyz** [comment]
@seriouszyx the demo looks good! Does it also support Keycloak SAML? We also need to use Casdoor as SP to connect to Keycloak SAML IDP.

File: web/src/App.js [hunk 1]
```diff
@@ -547,6 +549,7 @@ class App extends Component {
           <Route exact path="/signup/oauth/authorize" render={(props) => <LoginPage account={this.state.account} type={"code"} mode={"signup"} {...props} onUpdateAccount={(account) => {this.onUpdateAccount(account)}} />}/>
           <Route exact path="/login/oauth/authorize" render={(props) => <LoginPage account={this.state.account} type={"code"} mode={"signin"} {...props} onUpdateAccount={(account) => {this.onUpdateAccount(account)}} />}/>
           <Route exact path="/callback" component={AuthCallback}/>
+          <Route exact path="/saml/callback" component={SamlCallback}/>
```
Comment (hsluoyz) [inline web/src/App.js:552]: `/callback/saml` would be better?

File: web/src/App.js [hunk 2]
```diff
@@ -534,7 +535,8 @@ class App extends Component {
       window.location.pathname.startsWith("/login") ||
       window.location.pathname.startsWith("/callback") ||
       window.location.pathname.startsWith("/prompt") ||
-      window.location.pathname.startsWith("/forget");
+      window.location.pathname.startsWith("/forget") || 
+      window.location.pathname.startsWith("/saml/callback");
```
Comment (hsluoyz) [inline web/src/App.js:539]: No need for this line if it's `/callback/saml`

File: object/saml.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		panic(err)
+	}
+	return assertionInfo.NameID
+}
+
+func GenerateSamlLoginUrl(id string) string {
+	certStore := dsig.MemoryX509CertificateStore{
+		Roots: []*x509.Certificate{},
+	}
+	provider := GetProvider(id)
```
Comment (hsluoyz) [inline object/saml.go:70]: check this is a SAML provider here

File: object/saml.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	}
+	idpCert, err := x509.ParseCertificate(certData)
+	if err != nil {
+		panic(err)
+	}
+	certStore.Roots = append(certStore.Roots, idpCert)
+
+	samlOrigin := beego.AppConfig.String("samlOrigin")
+	sp := &saml2.SAMLServiceProvider{
+		ServiceProviderIssuer:       samlOrigin + "/api/acs",
```
Comment (hsluoyz) [inline object/saml.go:55]: Use `fmt.Sprintf` anywhere if possible, avoid string add

File: object/saml.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	if err != nil {
+		panic(err)
+	}
+	certStore.Roots = append(certStore.Roots, idpCert)
+	randomKeyStore := dsig.RandomKeyStoreForTest()
+	samlOrigin := beego.AppConfig.String("samlOrigin")
+	sp := &saml2.SAMLServiceProvider{
+		IdentityProviderSSOURL:      provider.Endpoint,
+		IdentityProviderIssuer:      provider.IssuerUrl,
+		ServiceProviderIssuer:       samlOrigin + "/api/acs",
```
Comment (hsluoyz) [inline object/saml.go:85]: Use `fmt.Sprintf` anywhere if possible, avoid string add

File: object/saml.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	"regexp"
+	"strings"
+
+	"github.com/astaxie/beego"
+	saml2 "github.com/russellhaering/gosaml2"
+	dsig "github.com/russellhaering/goxmldsig"
+)
+
+func ParseSamlResponse(samlResponse string) string {
+	certStore := dsig.MemoryX509CertificateStore{
```
Comment (hsluoyz) [inline object/saml.go:30]: Can we remove code duplicates with `GenerateSamlLoginUrl()`?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/auth/AuthBackend.js [hunk 7]
```diff
@@ -76,3 +76,18 @@ export function unlink(values) {
     body: JSON.stringify(values),
   }).then(res => res.json());
 }
+
+export function getSamlLogin() {
+  return fetch(`${authConfig.serverUrl}/api/get-saml-login?id=admin/aliyun-idaas`, {
```
Comment (Steve0x2a) [inline web/src/auth/AuthBackend.js:81]: Maybe it would be better to pass in idp's name

File: object/saml.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+		Roots: []*x509.Certificate{},
+	}
+	samlResponse, _ = url.QueryUnescape(samlResponse)
+	de, err := base64.StdEncoding.DecodeString(samlResponse)
+	if err != nil {
+		panic(err)
+	}
+	deStr := strings.Replace(string(de), "\n", "", -1)
+	res := regexp.MustCompile(`<ds:X509Certificate>(.*?)</ds:X509Certificate>`).FindAllStringSubmatch(deStr, -1)
+	str := res[0][0]
```
Comment (Steve0x2a) [inline object/saml.go:40]: idp may pass in multiple public keys

File: object/saml.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+	}
+	certStore.Roots = append(certStore.Roots, idpCert)
+	randomKeyStore := dsig.RandomKeyStoreForTest()
+	samlOrigin := beego.AppConfig.String("samlOrigin")
+	sp := &saml2.SAMLServiceProvider{
+		IdentityProviderSSOURL:      provider.Endpoint,
+		IdentityProviderIssuer:      provider.IssuerUrl,
+		ServiceProviderIssuer:       samlOrigin + "/api/acs",
+		AssertionConsumerServiceURL: samlOrigin + "/api/acs",
+		SignAuthnRequests:           false,
```
Comment (Steve0x2a) [inline object/saml.go:87]: We should implement authnrequest

**Review by Steve0x2a — CHANGES_REQUESTED** [review]
Good job, But we may need to make some improvements.

File: object/saml.go (hunk 8 above)
Comment (seriouszyx) [reply object/saml.go]: Yes, the metadata of Aliyun IDaaS is as below and you can find that the values of two `ds:X509Certificate` are the same. So `res[0][1]` and `res[0][0]` both work well.

```xml
<md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" entityID="CASDOOR-TEST">
<md:IDPSSODescriptor WantAuthnRequestsSigned="false" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
<md:KeyDescriptor use="signing">
<ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
<ds:X509Data>
<ds:X509Certificate>MIIB5TCCAU6gAwIBAgIIRiyDuM0koEswDQYJKoZIhvcNAQEFBQAwNTELMAkGA1UEBhMCQ04xDzANBgNVBAgMBuWMl+S6rDEVMBMGA1UEAxMMQ0FTRE9PUi1URVNUMB4XDTIxMTEyOTEwMTIwMloXDTI0MTEyODEwMTIwMlowNTELMAkGA1UEBhMCQ04xDzANBgNVBAgMBuWMl+S6rDEVMBMGA1UEAxMMQ0FTRE9PUi1URVNUMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCKxhfTOmQWs7i5dudkFxOHktushcrUloteFAYsKFhmlqtmc6M6upUAr59Yw14f+6myql/pWRdJIHMgJQuEcWF01aj4uHmwRrFpJHT+AbMBxfW8VNGl6JopHTLirAaTDESLhNaYS1YWxPluHRJ+7k3OE02ixQG5GLlf6EH1gm/8/QIDAQABMA0GCSqGSIb3DQEBBQUAA4GBADg4loZ7nAkDJqVSQMY8R/IiXemDzbECssSQBtLQLySzM0zkYij07AOTHOVVD6qQs4kULK5zQGWWDywjLFtitnwRWZ/DE3X7dJ5Vo2SwsPNtMQ2GKKXpWsurzUBzA8zjaFjwA1aJEXY/0tNkQuSs/pyNjoNtIq61zdmMdicpwB8X</ds:X509Certificate>
</ds:X509Data>
</ds:KeyInfo>
</md:KeyDescriptor>
<md:KeyDescriptor use="encryption">
<ds:KeyInfo xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
<ds:X509Data>
<ds:X509Certificate>MIIB5TCCAU6gAwIBAgIIRiyDuM0koEswDQYJKoZIhvcNAQEFBQAwNTELMAkGA1UEBhMCQ04xDzANBgNVBAgMBuWMl+S6rDEVMBMGA1UEAxMMQ0FTRE9PUi1URVNUMB4XDTIxMTEyOTEwMTIwMloXDTI0MTEyODEwMTIwMlowNTELMAkGA1UEBhMCQ04xDzANBgNVBAgMBuWMl+S6rDEVMBMGA1UEAxMMQ0FTRE9PUi1URVNUMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCKxhfTOmQWs7i5dudkFxOHktushcrUloteFAYsKFhmlqtmc6M6upUAr59Yw14f+6myql/pWRdJIHMgJQuEcWF01aj4uHmwRrFpJHT+AbMBxfW8VNGl6JopHTLirAaTDESLhNaYS1YWxPluHRJ+7k3OE02ixQG5GLlf6EH1gm/8/QIDAQABMA0GCSqGSIb3DQEBBQUAA4GBADg4loZ7nAkDJqVSQMY8R/IiXemDzbECssSQBtLQLySzM0zkYij07AOTHOVVD6qQs4kULK5zQGWWDywjLFtitnwRWZ/DE3X7dJ5Vo2SwsPNtMQ2GKKXpWsurzUBzA8zjaFjwA1aJEXY/0tNkQuSs/pyNjoNtIq61zdmMdicpwB8X</ds:X509Certificate>
</ds:X509Data>
</ds:KeyInfo>
</md:KeyDescriptor>
<md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://geonmahrnw.login.aliyunidaas.com/enduser/api/application/plugin_saml/idaas-cn-beijing-mu34ajc7ktlplugin_saml/sp_sso"/>
<md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://geonmahrnw.login.aliyunidaas.com/enduser/api/application/plugin_saml/idaas-cn-beijing-mu34ajc7ktlplugin_saml/sp_sso_post"/>
</md:IDPSSODescriptor>
</md:EntityDescriptor>
```

**Comment by seriouszyx** [comment]
@hsluoyz I refactored codes and added inputs to parse metadata automatically and showed SP ACS Url, SP Entity ID to copy.

The [demo link](http://82.156.216.21:8000/) and username/password are not changed. 

image

**Comment by hsluoyz** [comment]
@seriouszyx lgtm!

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**